pkg/controller/storageversionmigrator/storageversionmigrator.go

```diff
@@ -391,6 +415,78 @@ func (svmc *SVMController) failMigration(ctx context.Context, toBeProcessedSVM *
 	return errStatus
 }
 
+func (svmc *SVMController) beginCRDMigration(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) error {
+	cond := apihelpers.FindCRDCondition(crd, apiextensionsv1.StorageMigrating)
+	if cond != nil && cond.Status == apiextensionsv1.ConditionTrue {
+		return nil
+	}
+	// Set a migration running condition prior to running the migration.
+	cond = &apiextensionsv1.CustomResourceDefinitionCondition{
+		Type:               apiextensionsv1.StorageMigrating,
+		Status:             apiextensionsv1.ConditionTrue,
+		Reason:             "StorageMigration",
+		Message:            "Storage migration is currently in progress",
+		ObservedGeneration: crd.Generation,
+	}
+	apihelpers.SetCRDCondition(crd, *cond)
+	_, err := svmc.crdClient.UpdateStatus(ctx, crd, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (svmc *SVMController) finishCRDMigration(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) error {
+	crd, err := svmc.crdClient.Get(ctx, crd.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	cond := apihelpers.FindCRDCondition(crd, apiextensionsv1.StorageMigrating)
+	// Condition either unset or not in progress
+	if cond != nil && cond.Status != apiextensionsv1.ConditionTrue {
+		// This is an error but we should fail gracefully
+		return nil
+	}
+	oldGen := cond.ObservedGeneration
+	// Set a migration running condition prior to running the migration.
+	cond = &apiextensionsv1.CustomResourceDefinitionCondition{
+		Type:               apiextensionsv1.StorageMigrating,
+		Status:             apiextensionsv1.ConditionFalse,
+		Reason:             "StorageMigration",
+		Message:            "Storage migration is complete",
+		ObservedGeneration: crd.Generation,
+	}
+	// Wipe out all stored versions that we have migrated off of if nothing else
+	// has transpired during the migration.
+	if oldGen == crd.Generation {
+		var storedVersion string
+		for _, version := range crd.Spec.Versions {
+			if version.Storage {
+				storedVersion = version.Name
+			}
+		}
+		crd.Status.StoredVersions = []string{storedVersion}
+	}
+	apihelpers.SetCRDCondition(crd, *cond)
+	_, err = svmc.crdClient.UpdateStatus(ctx, crd, metav1.UpdateOptions{})
```

Comment:
is this update correctly preconditioned on the state of the CRD we based our decision on? if the CRD changed concurrently, will this fail with a conflict error? if so, do we need to put this entire chunk from the Get to the UpdateStatus in a RetryOnConflict block?

Reply:
Yeah since at this point we can be assured that the migration has completed successfully since the generation never changed until after all objects were migrated. Let me add a retry on conflict here. Worst case scenario is it fails, retries and the generation was incremented in the meantime making it so we don't know if it's safe to wipe the status.

Reply:
Added a retry block.